ducc/webhook/registry-listener.go

```diff
@@ -0,0 +1,77 @@
+package main
+
+import (
+  "bufio"
+  "fmt"
+  "io"
+  "os"
+  "log"
+  "time"
+  "flag"
+  "strings"
+  "os/exec"
+)
+
+func ReadChanges(file *os.File) chan string {
+
+    changes := make(chan string)
+
+    file.Seek(0, os.SEEK_END)
+    bf := bufio.NewReader(file)
+
+    go func() {
+        for {
+            line, _, err := bf.ReadLine()
+            if len(line) != 0 {
+                changes <- string(line)
+            } else if err == io.EOF {
+                time.Sleep(1 * time.Millisecond)
+            }
+        }
+    }()
+    return changes
+}
+
+func NotifyDucc(file_name string, repository_name string) {
+
+    file, err := os.OpenFile(file_name, os.O_RDONLY, 0755)
+    if err != nil {
+        log.Fatalf("OpenFile: %s", err)
+        }
+
+    changes := ReadChanges(file)
+
+    for {
+        msg := <-changes
+
+        if msg == "xx|file rotation|xx" {
+                NotifyDucc(file_name, repository_name)
```

Comment:
This recursive call means that this long-running process is leaking file descriptors. The file opened inside `ReadChanges` is opened but never closed. It also means the old (pre-rotation) version of the file is kept around until the process is finally stopped.